### PR TITLE
[WIP] SpatialVector, SpatialVelocity, and SpatialInertia from Jain's textbook

### PR DESCRIPTION
**Pull request by angadhn**

### CODE REVIEW

**Comment by angadhn** [comment]
@moorepants

**Comment by moorepants** [comment]
Cool, I will try to look at it by tomorrow to give some comments.

**Comment by moorepants** [comment]
I haven't looked at the PR in great detail, but wanted to mention one critical aspect, in my opinion. The physics vector.py is designed around basis independent vectors and dyadics. This is certainly an influence of  the Kane view on dynamics. You current design is a basis dependent design and does not utilize any of the current reference frame machinery. Ideally I would like to see basis independent spatial vectors, etc that have the same api and display style as what we already have the in mechanics module. A first class SpatialVector would feel like the current vector allowing them to be expressed in unit vectors from any reference frame and all of the operations on these vectors would then follow the rules of spatial vector operators.

We can implement a basis dependent module for doing dynamics with spatial quantities, but we should be careful how it is mixed with the current system. The attempt at this two years ago also did not implement a first class basis-independent spatial vector and I had the same feedback then.

What are the pros and cons of not implementing the spatial vector in the same style as our current vector?

**Comment by angadhn** [comment]
To start I will take the quote from Abhi's book "Robot and Multibody Dynamics":
"While coordinate-free notation is simpler, actual computations require working with coordinate-frame representations of vectors."

My initial design began with what you have suggested: a basis-independent spatial vector implementation which would return purely a 6-d column vector. However, it is insufficient as soon as we want to perform a simple task even for single body kinematics (i.e. 2 point theorem) using spatial operators (section 1.4.1 of Jain book). This requires a rotation matrix to generate the Phi and PhiStar matrices which are the subject of section 1.4 of Abhi's book. The RotationMatrix function is essentially just the ReferenceFrame's dcm operation; while the operation is redundant, having a function with this name just makes sense to me because people are so used to this name.

Also, I disagree that I do not use the existing architecture. ReferenceFrame and Vector classes are extremely integral to what is presented here. I also make use of the body class's use of the frame object; all of this is being done in a manner that maintains consistency with Abhi's textbook. Also, calling a SpatialVelocity instance is a tuple that provides of (Vector6, ReferenceFrame).

Of course, one could choose to not implement a 2PtTheorem for SpatialVelocity and just work with the equivalent Vector class operations but I think this will be limiting in the long run. The phi and phiStar matrices are not only useful for computing velocities but I'm pretty certain (based on earlier conversations with Abhi) that they are integral to generating Jacobians and also performing the O(n) mass matrix inversion shown in the dynamics equations; I should have more on the latter by next week and will likely have another PR on serial chain kinematics soon, as well.

To summarize, I tried coordinate-free representation but from another read of Abhi's book it is clear that I need to keep tabs on the ReferenceFrame object for computation of the symbolic equations even at the kinematic level. Hence, the clear mention of the ReferenceFrame even in the SpatialVector's call function. I am trying to stick very closely to Abhi's book; it is essentially what is in DARTS (hence the name of the PR's branch as PyDARTS. But maybe we can go with something catchier which I can tell you in chat :) It's THAT good! )

**Comment by czgdp1807** [comment]
@moorepants Can we close this? This PR isn't in a good state IMO.
